Engine_Revit_UI/Query/CeilingPattern.cs

```diff
@@ -0,0 +1,142 @@
+#else
+                fillPatternElement = revitMaterial.Document.GetElement(revitMaterial.SurfacePatternId) as FillPatternElement;
+#endif
+
+                if (fillPatternElement != null)
+                {
+                    FillPattern fillPattern = fillPatternElement.GetFillPattern();
+                    if (fillPattern == null || fillPattern.IsSolidFill)
+                        continue; //Skip solid filled patterns
+
+                    IList<FillGrid> fillGridList = fillPattern.GetFillGrids();
+                    foreach(FillGrid grid in fillGridList)
+                    {
+                        double offset = grid.Offset.ToSI(UnitType.UT_Length);
+
+                        double currentY = lowestY - (yLength / 2);
+
+                        while((currentY + offset) < (highestY + (yLength + 2)))
+                        {
+                            BH.oM.Geometry.Point pt = new oM.Geometry.Point { X = lowestX, Y = currentY + offset, Z = z };
+                            BH.oM.Geometry.Point pt2 = new oM.Geometry.Point { X = highestX, Y = currentY + offset, Z = z };
+
+                            BH.oM.Geometry.Line pline = new oM.Geometry.Line { Start = pt, End = pt2 };
+
+                            if (grid.Angle > 0)
+                            {
+                                BH.oM.Geometry.Point rotatePt = new BH.oM.Geometry.Line { Start = pt, End = pt2 }.Centroid();
+                                pline = pline.Rotate(rotatePt, Vector.ZAxis, grid.Angle.ToSI(UnitType.UT_Angle));
+
+                                BH.oM.Geometry.Point intersect = pline.LineIntersections(leftLine, true).FirstOrDefault();
+                                if(intersect != null)
+                                    pline = pline.Extend(pline.Start.Distance(intersect));
+
+                                intersect = pline.LineIntersections(rightLine, true).FirstOrDefault();
+                                if (intersect != null)
+                                    pline = pline.Extend(0, pline.End.Distance(intersect));
+                            }
+
+                            List<BH.oM.Geometry.Point> intersections = new List<oM.Geometry.Point>();
+                            foreach(BH.oM.Geometry.Line l in boundarySegments)
+                            {
+                                BH.oM.Geometry.Point p = pline.LineIntersection(l);
+                                if (p != null)
+                                    intersections.Add(p);
+                            }
+
+                            if(intersections.Count > 0)
+                            {
+                                List<BH.oM.Geometry.Line> lines = pline.SplitAtPoints(intersections);
+                                foreach(BH.oM.Geometry.Line l in lines)
+                                {
+                                    if (surface.ExternalBoundary.IIsContaining(l.IControlPoints(), true))
```

Comment:
Cosmetics, but here you could take midpoint of a curve instead of its both endpoints - could save 50% of calls to heavy `IsContaining`

Reply:
Opting not to resolve this one with a code change as I did toy with a midpoint of the curve but found some did occasionally get a false positive, so opted to check both points of a line for safety/accuracy at the expense of speed.

Reply:
Hmm I am trying to come up with a case on which it broke and cannot think of one. Do you remember when did that happen?